### DISCUSSION TITLE
Any existing RSA circuit implementation?

### DISCUSSION THREAD

**Discussion by doutv** [Consensys/gnark#1199] [Q&A] [answered]
I cannot find existing RSA implementation in Gnark. 

Could someone give advice on how to implement a new one?

**Comment by ivokub** [comment, marked as answer]
Field emulation allows to perform modular exponentiation for a modulus defined as witness. I think this should be sufficient for implementing RSA verification?

See [Field.ModExp](https://pkg.go.dev/github.com/consensys/gnark@v0.10.0/std/math/emulated#Field.ModExp) and [test implementation](https://github.com/Consensys/gnark/blob/v0.10.0/std/math/emulated/custommod_test.go#L139-L171).

Note that checking RSA signatures may also require doing padding computation a la RSA-OAEP or RSA-PSS. We don't have the padding implementation and I think we're not planning to implement ourselves in gnark as RSA is generally being deprecated in favor of ECC based signature schemes (Schnorr, EdDSA, ECDSA) or post-quantum secure schemes. And RSA verification for reasonable key sizes (3k+) in circuit is very inefficient.
